physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,583 @@
+
+// Delegate is an implementation of autopilot.ApplicationIntegration interface.
+// This is used by the autopilot library to retrieve information and to have
+// application specific tasks performed.
+type Delegate struct {
+	*RaftBackend
+}
+
+// AutopilotConfig is called by the autopilot library to know the desired
+// autopilot configuration.
+func (d *Delegate) AutopilotConfig() *autopilot.Config {
+	d.l.RLock()
+	config := &autopilot.Config{
+		CleanupDeadServers:      false,
+		LastContactThreshold:    d.autopilotConfig.LastContactThreshold,
+		MaxTrailingLogs:         d.autopilotConfig.MaxTrailingLogs,
+		MinQuorum:               d.autopilotConfig.MinQuorum,
+		ServerStabilizationTime: d.autopilotConfig.ServerStabilizationTime,
+		Ext:                     d.autopilotConfigExt(),
+	}
+	d.l.RUnlock()
+	return config
+}
+
+// NotifyState is called by the autopilot library whenever there is a state
+// change. We update a few metrics when this happens.
+func (d *Delegate) NotifyState(state *autopilot.State) {
+	if d.raft.State() == raft.Leader {
+		metrics.SetGauge([]string{"autopilot", "failure_tolerance"}, float32(state.FailureTolerance))
+		if state.Healthy {
+			metrics.SetGauge([]string{"autopilot", "healthy"}, 1)
+		} else {
+			metrics.SetGauge([]string{"autopilot", "healthy"}, 0)
+		}
+
+		for id, state := range state.Servers {
+			if state.Health.Healthy {
+				metrics.SetGauge([]string{"autopilot", string(id), "healthy"}, 1)
+			} else {
+				metrics.SetGauge([]string{"autopilot", string(id), "healthy"}, 0)
+			}
+		}
+	}
+}
+
+// FetchServerStats is called by the autopilot library to retrieve information
+// about all the nodes in the raft cluster.
+func (d *Delegate) FetchServerStats(ctx context.Context, servers map[raft.ServerID]*autopilot.Server) map[raft.ServerID]*autopilot.ServerStats {
+	ret := make(map[raft.ServerID]*autopilot.ServerStats)
+
+	d.l.RLock()
+	followerStates := d.followerStates
+	d.l.RUnlock()
+
+	followerStates.l.RLock()
+	defer followerStates.l.RUnlock()
+
+	now := time.Now()
+	for id, followerState := range followerStates.followers {
+		ret[raft.ServerID(id)] = &autopilot.ServerStats{
+			LastContact: now.Sub(followerState.LastHeartbeat),
+			LastTerm:    followerState.LastTerm,
+			LastIndex:   followerState.AppliedIndex,
+		}
+	}
+
+	leaderState, _ := d.fsm.LatestState()
+	ret[raft.ServerID(d.localID)] = &autopilot.ServerStats{
+		LastTerm:  leaderState.Term,
+		LastIndex: leaderState.Index,
+	}
+
+	return ret
+}
+
+// KnownServers is called by the autopilot library to know the status of each
+// node in the raft cluster. If the application thinks that certain nodes left,
+// it is here that we let the autopilot library know of the same.
+func (d *Delegate) KnownServers() map[raft.ServerID]*autopilot.Server {
+	d.l.RLock()
+	future := d.raft.GetConfiguration()
+	if err := future.Error(); err != nil {
+		d.logger.Error("failed to get raft configuration when computing known servers", "error", err)
```

Comment:
Might be best to unlock and return nil in this case.  Though looking at callers it seems they too call GetConfiguration and return before calling us on failure, so it's probably moot.

Reply:
I agree. Done.